[JIT] Load original SourceRanges on import

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22180 [JIT] Load original SourceRanges on import**
* #22179 [JIT] Write out SourceRanges on save()
* #22178 [JIT] Serialization methods for SourceRange and Source
* #22177 [JIT] Refactor file:line:col to be less ugly

Resolves https://github.com/pytorch/pytorch/issues/21183

Demo:

```
import torch
import torch.nn as nn

class M(torch.jit.ScriptModule):
    def __init__(self):
        super(M, self).__init__()
        self.conv = nn.Conv2d(5, 5, 2)

    @torch.jit.script_method
    def forward(self, x):
        return self.conv(x)

M().save("m.pt")
loaded = torch.jit.load("m.pt")
loaded(torch.ones(20, 20))
```

```
Traceback (most recent call last):
  File "demo.py", line 15, in <module>
    loaded(torch.ones(20, 20))
  File "/home/jamesreed/pytorch/torch/nn/modules/module.py", line 531, in __call__
    result = self.forward(*input, **kwargs)
RuntimeError: Expected 4-dimensional input for 4-dimensional weight 5 5, but got 2-dimensional input of size [20, 20] instead
The above operation failed in interpreter, with the following stack trace:
at code/m.py:248:10
            ret10 = x
          ret5 = ret10
        ret0 = ret5
      ret = ret0
    _122 = _0.weight
    _123 = _0.bias
    _124 = [0, 0]
    _1 = torch.conv2d(ret, _122, _123, [1, 1], _124, [1, 1], 1)
  else:
    _1 = torch.conv2d(x, _0.weight, _0.bias, [1, 1], [0, 0], [1, 1], 1)
         ~~~~~~~~~~~~ <--- HERE
  return _1
Compiled from code at /home/jamesreed/pytorch/torch/nn/modules/conv.py:344:16
    @weak_script_method
    def forward(self, input):
        if self.padding_mode == 'circular':
            expanded_padding = ((self.padding[1] + 1) // 2, self.padding[1] // 2,
                                (self.padding[0] + 1) // 2, self.padding[0] // 2)
            return F.conv2d(F.pad(input, expanded_padding, mode='circular'),
                            self.weight, self.bias, self.stride,
                            _pair(0), self.dilation, self.groups)
        return F.conv2d(input, self.weight, self.bias, self.stride,
               ~~~~~~~~ <--- HERE
                        self.padding, self.dilation, self.groups)
```


Differential Revision: [D15981425](https://our.internmc.facebook.com/intern/diff/D15981425)

Commit:
[JIT] Load original SourceRanges on import

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1118,9 +1123,15 @@ struct PythonPrintPass {
-    SourceRangeCtxMgr guard(
-        &source_range_stack_,
-        std::make_shared<SourceRange>(graph.param_node()->sourceRange()));
+    std::shared_ptr<SourceRange> debug_range;
```
No copy pasta, just make SourceRangeCtxMgr do this.

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -9,25 +9,65 @@
 namespace jit {
 
+struct SourceRange;
+using SourceRangeRecord = std::tuple<size_t, std::shared_ptr<SourceRange>>;
```
please use named classes. This stuff is always confusing to read.

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -9,25 +9,65 @@
+using SourceRangeRecord = std::tuple<size_t, std::shared_ptr<SourceRange>>;
+using SourceRangeRecords = std::vector<SourceRangeRecord>;
+
+// Class that keeps track of a serialized debug info table and lazily
```
This comment is not useful. I don't know what is in data using this comment.

zdevito replied:
This is implementation detail and doesn't need to be in source_range.h, which is widely included. It can be opaque.

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -9,25 +9,65 @@
-      : text_(std::move(text)), filename_(c10::nullopt) {
+  explicit Source(
+      std::string text,
+      std::shared_ptr<DebugInfo> debug_info = nullptr)
```
DebugInfo is not a great name here because it doesn't really describe what it does. After all, all of SourceRange/Source is really 'debug info'. It would be better if this were a chain. i.e. generating_source, the source that generated this source.

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -79,6 +119,25 @@ struct Source {
+        std::move(text_), std::move(filename_), starting_line_no_);
+  }
+
+  std::shared_ptr<SourceRange> query_debug_info(const SourceRange& sr) const {
```
The name of this function doesn't make sense to me. It gives me no idea what it is actually doing. Maybe something like: `findSourceRangeThatGenerated(`

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -170,6 +231,26 @@ struct CAFFE2_API SourceRange {
+  }
+
+  std::shared_ptr<SourceRange> orig_range() const {
+    // Cache this to prevent log(n) lookup every time.
```
SourceRange objects are value objects. They are passed by value all over the place in the compiler. Trying to cache the lookup here won't work because there won't be cases where two things share the same source range. I think it would be possible to keep the cache inside the DebugInfo object, however, where it can keep a hash_table of queried byte offset -> SourceRange.

zdevito reviewed (commented):
Cool -- see inline comments for some suggestions. Need to be careful about SourceRanges continuing to be value types (caching has to go at the Source level if at all). Also, I didn't immediately see it in the code, but what happens if you query a source range where the originating code comes from many different locations? Is it returning the source location at the beginning of the source range?

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[WIP][JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[WIP][JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[WIP][JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1238,8 +1247,13 @@ struct PythonPrintPass {
+    out << getImports();
+    int64_t source_offset = out.tellp();
+    out << body_.str();
+    for (const auto& x : body_.ranges()) {
```
I know this is small but this repeating the internal details of source range tracking outside of the the TaggedStream object. This will be a nasty and hard-to-find bug if someone changes the representation and doesn't realize the logic is also repeated here.

zdevito commented on torch/csrc/jit/source_range_serialization.cpp:
```diff
@@ -0,0 +1,135 @@
+
+ private:
+  std::shared_ptr<Source> deserialize_source(const c10::IValue& iv) {
+    // TODO: cache these?
```
Yeah, we need to do this one because otherwise it is copy the entire source string over and over again.

zdevito replied:
The key is just the pointer that is being stored inside in the IValue. I think we have a method that accesses it (see the pickler, which does something similar).

zdevito reviewed (approved):
Looks good! Combined with @ZolotukhinM's  inlining remembering stack changes this will enable a full description of the stack. Keep in mind that We are going to end up in a situation where there are two separate traces: the one from the loaded code and the original one. Their depth is not the same and we need to do a good job reporting it all. We should resolve this when we merge the components together.

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head

Commit:
Update on "[JIT] Load original SourceRanges on import"

[JIT] Load original SourceRanges on import

gh-metadata: pytorch pytorch 22180 gh/jamesr66a/9/head
